lib/pbench/server/database/models/audit.py

```diff
@@ -0,0 +1,370 @@
+    AuditReason and possibly further detailed by a message."""
+    FAILURE = enum.auto()
+
+    """Warning: the operation didn't fail, but wasn't able to complete with
+    total success."""
+    WARNING = enum.auto()
+
+
+class AuditReason(enum.Enum):
+
+    """Permission denied."""
+
+    PERMISSION = enum.auto()
+
+    """Internal failure."""
+    INTERNAL = enum.auto()
+
+    """Consistency failure. For example, a corrupted tarball upload."""
+    CONSISTENCY = enum.auto()
+
+
+class Audit(Database.Base):
+    """
+    A framework to store Pbench audit records. These will track server
+    configuration changes as well as every mutation of user-visible data and
+    attribute each to a specific user and operation.
+
+    Architecturally, the SQL table may be used as a "front end" for a more
+    expensive audit repository (e.g., Elasticsearch). For example, we might
+    periodically bulk-replicate these rows as JSON data into an Elasticsearch
+    index and wipe the table.
+
+    Columns:
+        id: Generated unique ID of table row
+        root_id: Reference to the id of the operation's BEGIN audit record to
+            enable following an operation with multiple records.
+        name: Provide an optional name for the operation, like "upload" or
+            "index".
+        operation: The operation performed
+        object_type: The type of resource
+        object_id: The object on which an operation was performed
+        object_name: The name of the object (more meaningful than ID if the
+            resource has been deleted)
+        user_id: The identity of the user performing an operation
+        user_name: The name of the user (more meaningful than ID if the
+            user has been deleted)
+        status: The status of the operation
+        attributes: A JSON document of attributes ("message" key is a common
+            convention but others may be included for clarity)
+        timestamp: The date/time the operation was performed
+    """
+
+    __tablename__ = "audit"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    root_id = Column(Integer, nullable=True)
+    name = Column(String(128), nullable=True)
+    operation = Column(Enum(OperationCode), index=True, nullable=False)
+    object_type = Column(Enum(AuditType), index=False, nullable=True)
+    object_id = Column(String(128), index=False, nullable=True)
+    object_name = Column(String(256), nullable=True)
+    user_id = Column(String(128), index=True, nullable=True)
+    user_name = Column(String(256), index=False, nullable=True)
+    status = Column(Enum(AuditStatus), index=True, nullable=False)
+    reason = Column(Enum(AuditReason), nullable=True)
+    attributes = Column(JSON, index=False, nullable=True)
+    timestamp = Column(TZDateTime, nullable=False, default=TZDateTime.current_time)
+
+    # Other global class properties
+
+    BACKGROUND_USER = "BACKGROUND"  # Operation performed in background
```

Comment:
But, couldn't we have multiple indexers?  (Perhaps I'm conflating unpacking with indexing, but it seems like we are likely to have some parallelism in our background activities, so we might want to be able to differentiate between actors when things are running concurrently.)

Reply:
I had considered tracking *pid*, which would let us disambiguate which of two parallel indexers did the deed. But unless we have a persistent component that runs a long time processing sequential datasets, the pid is pretty transient and difficult to trace. Of course, if we have an opportunistic lambda that processes a second dataset if the message is available instead of terminating, there may be some value in knowing they were generated by the same process. Then again, tracking pids across multiple containers without also somehow tracking container/pod identification probably doesn't help much. I decided not to try to get into any of this. We can always add it later.